_posts/2020-12-21--kinesis-streams-part-2.md

```diff
@@ -0,0 +1,591 @@
+
+Though the direct integrations with _EMR_, _Kinesis Firehose_, or _Analytics_ can only use the shared throughput, the great news is, you can use enhanced fan-out with your custom consumers. In consumer applications using KCL enabling enhanced fan-out is just a matter of configuration. With both, custom consumers using AWS SDK and _AWS Lambda_, you need to create a so-called **enhanced stream consumer** first.
+
+I know, naming gets confusing here, but bear with me...
+
+
+![highway](/img/kinesis/highway.jpg){: .img.centered }
+
+
+You can think of an enhanced consumer for Lambda (or SDK) applications simply as of an **isolated connection** to your stream, sort of a private highway of yours. All it takes to create such an enhanced consumer is a simple API call. You could also use e.g. _CloudFormation_ or _SAM_, or you can just run this simple _AWS CLI_ command: 
+
+
+```
+aws kinesis register-stream-consumer --consumer-name my_highway \
+      --stream-arn arn:aws:kinesis:<region>:<account_id>:stream/the-best-stream
+```
+
+Though apparently, there's no way to register an enhanced consumer through the _AWS Kinesis_ console yet (if you find one, let me know!), you can list and delete the existing ones:
+
+
+![consumer_console](/img/kinesis/consumer_console.png){: .img.centered }
+
+ 
+
+In the case of _AWS Lambda_, you can now use this enhanced consumer when configuring the Kinesis trigger for your function. Note, that you can only have **a single Lambda function** using a given enhanced consumer. You will need to create a separate enhanced stream consumer for each of your Lambda functions.
+
+
+
+### Which one to chose?
+
+Though the dedicated throughput sounds great on its own, this is not the only perk you are getting with enhanced fan-out. Together with it comes even less **read propagation latency**, aka the time between when the data is written to the stream and when it is consumed. This is in part due to the dedicated throughput “highway" each consumer is getting, and in part to the fact that enhanced fan-out uses **HTTP2** instead of HTTP, allowing to use persistent connections and to **push** records to the consumers, instead of consumers **polling** the stream.
+
+When using enhanced fan-out, you will have records available to be consumed about _70ms_ after you wrote them to the stream. In comparison, it takes about _200ms_ for the shared throughput consumers and this latency increases with each additional consumer that shares the throughput. Why is that? You will have to keep reading to find out 😉 
+
+By now, you might be thinking, why would anyone use the standard shared throughput at all? One important reason is this: while the standard shared throughput consumers normally read the data **for free**, you will have to [pay](https://aws.amazon.com/kinesis/data-streams/pricing/) for having all the perks of enhanced fan-out. You will pay for both, the **amount of data** consumed, as well as for the **number of consumers per shard**.
```

Comment:
I'd drop the commas: "By now you might be thinking why would anyone use the standard shared throughput at all?"